File: lib/handlers.ts
import { config } from "./config";
import { db } from "./db";
import { telegram } from "./telegram";
import { CONFIRM_PREFIX, REJECT_PREFIX, handleResponse } from "./queueManager";

interface TgUser {
  id: number;
  username?: string;
  first_name?: string;
  last_name?: string;
  is_bot?: boolean;
}

interface TgChat {
  id: number;
  type: string;
}

interface TgMessage {
  message_id: number;
  chat: TgChat;
  from?: TgUser;
  text?: string;
  new_chat_members?: TgUser[];
}

interface TgCallbackQuery {
  id: string;
  data?: string;
  from: TgUser;
  message?: TgMessage;
}

export interface TgUpdate {
  update_id: number;
  message?: TgMessage;
  callback_query?: TgCallbackQuery;
}

function fullName(user: TgUser): string {
  return [user.first_name, user.last_name].filter(Boolean).join(" ") || user.username || String(user.id);
}

function isAdmin(userId: number): boolean {
  return config.adminIds().has(String(userId));
}

function command(text: string | undefined): string | undefined {
  if (!text || !text.startsWith("/")) return undefined;
  return text.split(/[\s@]/)[0].slice(1).toLowerCase();
}

async function handleStart(message: TgMessage) {
  if (message.chat.type !== "private" || !message.from) {
    await telegram.sendMessage(
      message.chat.id,
      "Please message me privately with /start first so I can DM you job alerts."
    );
    return;
  }
  const user = message.from;
  await db.upsertMember(user.id, user.username ?? null, fullName(user));
  await db.setPrivateChat(user.id, message.chat.id);
  await telegram.sendMessage(
    message.chat.id,
    "You're registered. Go to the group and send /queue to join the job queue."
  );
}

async function handleNewMembers(message: TgMessage) {
  const members = (message.new_chat_members ?? []).filter((m) => !m.is_bot);
  for (const member of members) {
    await db.upsertMember(member.id, member.username ?? null, fullName(member));
  }
  if (members.length) {
    const names = members.map(fullName).join(", ");
    await telegram.sendMessage(
      message.chat.id,
      `Welcome ${names}! DM me /start, then send /queue here to join the job queue.`
    );
  }
}

async function handleJoinQueue(message: TgMessage) {
  if (!message.from) return;
  const user = message.from;
  await db.upsertMember(user.id, user.username ?? null, fullName(user));

  const member = await db.getMember(user.id);
  if (!member || !member.private_chat_id) {
    await telegram.sendMessage(
      message.chat.id,
      "Please DM me /start first so I'm able to send you job alerts privately."
    );
    return;
  }

  if (await db.isInQueue(user.id)) {
    const position = await db.getQueuePosition(user.id);
    await telegram.sendMessage(message.chat.id, `You're already in the queue at position ${position}.`);
    return;
  }

  await db.addToQueue(user.id);
  const position = await db.getQueuePosition(user.id);
  await telegram.sendMessage(message.chat.id, `You've joined the queue at position ${position}.`);
}

async function handleLeaveQueue(message: TgMessage) {
  if (!message.from) return;
  const user = message.from;
  if (!(await db.isInQueue(user.id))) {
    await telegram.sendMessage(message.chat.id, "You're not in the queue.");
    return;
  }
  await db.removeFromQueue(user.id);
  await telegram.sendMessage(message.chat.id, "You've left the queue.");
}

async function handlePosition(message: TgMessage) {
  if (!message.from) return;
  const position = await db.getQueuePosition(message.from.id);
  if (position === undefined) {
    await telegram.sendMessage(message.chat.id, "You're not in the queue. Send /queue to join.");
    return;
  }
  await telegram.sendMessage(message.chat.id, `You're at position ${position} in the queue.`);
}

async function handleQueueList(message: TgMessage) {
  const rows = await db.getQueue();
  if (!rows.length) {
    await telegram.sendMessage(message.chat.id, "The queue is empty.");
    return;
  }
  const lines = rows.map((row, i) => `${i + 1}. ${row.username ?? row.full_name}`);
  await telegram.sendMessage(message.chat.id, "Current queue:\n" + lines.join("\n"));
}

async function handleRecentJobs(message: TgMessage) {
  if (!message.from || !isAdmin(message.from.id)) {
    await telegram.sendMessage(message.chat.id, "Admins only.");
    return;
  }
  const rows = await db.listRecentJobs();
  if (!rows.length) {
    await telegram.sendMessage(message.chat.id, "No jobs recorded yet.");
    return;
  }
  const lines = rows.map((row) => `#${row.id} [${row.status}] ${row.subject}`);
  await telegram.sendMessage(message.chat.id, "Recent jobs:\n" + lines.join("\n"));
}

export async function handleUpdate(update: TgUpdate): Promise<void> {
  if (update.message) {
    const message = update.message;
    if (message.new_chat_members?.length) {
      await handleNewMembers(message);
      return;
    }
    switch (command(message.text)) {
      case "start":
        return handleStart(message);
      case "queue":
        return handleJoinQueue(message);
      case "leave":
        return handleLeaveQueue(message);
      case "position":
        return handlePosition(message);
      case "queuelist":
        return handleQueueList(message);
      case "jobs":
        return handleRecentJobs(message);
      default:
        return;
    }
  }

  if (update.callback_query) {
    const query = update.callback_query;
    const data = query.data ?? "";
    let assignmentId: number | undefined;
    let confirmed = false;

    if (data.startsWith(CONFIRM_PREFIX)) {
      assignmentId = Number(data.slice(CONFIRM_PREFIX.length));
      confirmed = true;
    } else if (data.startsWith(REJECT_PREFIX)) {
      assignmentId = Number(data.slice(REJECT_PREFIX.length));
      confirmed = false;
    } else {
      await telegram.answerCallbackQuery(query.id);
      return;
    }

    const reply = await handleResponse(assignmentId, confirmed);
    await telegram.answerCallbackQuery(query.id);
    if (query.message) {
      await telegram.editMessageReplyMarkup(query.message.chat.id, query.message.message_id);
      await telegram.sendMessage(query.message.chat.id, reply);
    }
  }
}

File: bot/handlers.py
from telegram import Update
from telegram.ext import ContextTypes

from . import config, database, queue_manager


def is_admin(user_id: int) -> bool:
    return user_id in config.ADMIN_IDS


async def start(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if update.effective_chat.type != "private":
        await update.message.reply_text(
            "Please message me privately with /start first so I can DM you job alerts."
        )
        return

    user = update.effective_user
    await database.upsert_member(user.id, user.username, user.full_name)
    await database.set_private_chat(user.id, update.effective_chat.id)
    await update.message.reply_text(
        "You're registered. Go to the group and send /queue to join the job queue."
    )


async def on_new_members(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    for member in update.message.new_chat_members:
        if member.is_bot:
            continue
        await database.upsert_member(member.id, member.username, member.full_name)
    names = ", ".join(m.full_name for m in update.message.new_chat_members if not m.is_bot)
    if names:
        await update.message.reply_text(
            f"Welcome {names}! DM me /start, then send /queue here to join the job queue."
        )


async def join_queue(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    user = update.effective_user
    await database.upsert_member(user.id, user.username, user.full_name)

    member = await database.get_member(user.id)
    if member is None or member["private_chat_id"] is None:
        await update.message.reply_text(
            "Please DM me /start first so I'm able to send you job alerts privately."
        )
        return

    if await database.is_in_queue(user.id):
        position = await database.get_queue_position(user.id)
        await update.message.reply_text(f"You're already in the queue at position {position}.")
        return

    await database.add_to_queue(user.id)
    position = await database.get_queue_position(user.id)
    await update.message.reply_text(f"You've joined the queue at position {position}.")


async def leave_queue(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    user = update.effective_user
    if not await database.is_in_queue(user.id):
        await update.message.reply_text("You're not in the queue.")
        return
    await database.remove_from_queue(user.id)
    await update.message.reply_text("You've left the queue.")


async def my_position(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    user = update.effective_user
    position = await database.get_queue_position(user.id)
    if position is None:
        await update.message.reply_text("You're not in the queue. Send /queue to join.")
        return
    await update.message.reply_text(f"You're at position {position} in the queue.")


async def list_queue(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    rows = await database.get_queue()
    if not rows:
        await update.message.reply_text("The queue is empty.")
        return
    lines = [
        f"{i}. {row['username'] or row['full_name']}" for i, row in enumerate(rows, start=1)
    ]
    await update.message.reply_text("Current queue:\n" + "\n".join(lines))


async def recent_jobs(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if not is_admin(update.effective_user.id):
        await update.message.reply_text("Admins only.")
        return
    rows = await database.list_recent_jobs()
    if not rows:
        await update.message.reply_text("No jobs recorded yet.")
        return
    lines = [f"#{row['id']} [{row['status']}] {row['subject']}" for row in rows]
    await update.message.reply_text("Recent jobs:\n" + "\n".join(lines))


async def on_callback(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    query = update.callback_query
    data = query.data or ""

    if data.startswith(queue_manager.CONFIRM_PREFIX):
        assignment_id = int(data[len(queue_manager.CONFIRM_PREFIX):])
        confirmed = True
    elif data.startswith(queue_manager.REJECT_PREFIX):
        assignment_id = int(data[len(queue_manager.REJECT_PREFIX):])
        confirmed = False
    else:
        await query.answer()
        return

    reply = await queue_manager.handle_response(context, assignment_id, confirmed)
    await query.answer()
    await query.edit_message_reply_markup(reply_markup=None)
    await query.message.reply_text(reply)
